### PR TITLE
Added template wizard and palette change

### PR DESCRIPTION
**Pull request by mibrah42**
## What are we adding in this PR

- [X] `template-wizard` widget

image

- [x] Change palette button for switching color palettes

_Changing palettes trigger a confirmation dialog to avoid accidentally overwriting template styles_

### CODE REVIEW

File: app-creator/client/src/widgets/dropzone-widget/dropzone-widget.ts
Comment (mibrah42) [inline app-creator/client/src/widgets/dropzone-widget/dropzone-widget.ts:141]: Resolves issue: https://github.com/googleinterns/earthengine-community/projects/1#card-42084226 @joannalcy

File: app-creator/client/src/widgets/story-board/story-board.ts
Comment (joannalcy) [inline app-creator/client/src/widgets/story-board/story-board.ts]: Uncompleted sentence? :)

File: app-creator/client/src/utils/helpers.ts
Comment (joannalcy) [inline app-creator/client/src/utils/helpers.ts:69]: Can this be replaced with `Array.fill()`?

File: app-creator/client/src/utils/palettes.ts
Comment (joannalcy) [inline app-creator/client/src/utils/palettes.ts]: Is it possible to turn this into a `Record`? See https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkt

File: app-creator/client/src/widgets/dropzone-widget/dropzone-widget.ts
Comment (joannalcy) [inline app-creator/client/src/widgets/dropzone-widget/dropzone-widget.ts]: Treat abbreviations like acronyms in names as whole words, i.e. use childrenIds, not childrenIDs, unless required by a platform name (e.g. XMLHttpRequest).

File: app-creator/client/src/widgets/dropzone-widget/dropzone-widget.ts
Comment (joannalcy) [inline app-creator/client/src/widgets/dropzone-widget/dropzone-widget.ts:145]: Is it possible to do the removal outside of the `getChildrenIds()` method? `getChildrenIds()` should ideally only return the IDs and not produce any side effects.

File: app-creator/client/src/widgets/story-board/story-board.ts
Comment (joannalcy) [inline app-creator/client/src/widgets/story-board/story-board.ts]: Can we use `clearSelectedPalette()` instead?

File: app-creator/client/src/utils/helpers.ts
Comment (mibrah42) [reply app-creator/client/src/utils/helpers.ts:69]: Array.fill will calculate this value `(~~(Math.random() * 36)).toString(36))` one time and populate all elements in the array with it (i.e. ["a", "a", "a", ...]).

File: app-creator/client/src/widgets/dropzone-widget/dropzone-widget.ts
Comment (mibrah42) [reply app-creator/client/src/widgets/dropzone-widget/dropzone-widget.ts:145]: The `getChildrenIds` method only has this, everything else is outside.

```Typescript
 /**
   * Return children IDs for a given dropzone.
   */
  getChildrenIds(): string[] {
    return Array.from(this.children).map((el) => {
      return el.firstElementChild!.id;
    });
  }
```

Comment (mibrah42) [reply app-creator/client/src/widgets/dropzone-widget/dropzone-widget.ts:145]: I think it looked like its all inside because the code between was collapsed

Screenshot 2020-08-03 at 5 26 15 PM

File: app-creator/client/src/widgets/story-board/story-board.ts
Comment (mibrah42) [reply app-creator/client/src/widgets/story-board/story-board.ts]: ClearSelectedPalette reverts to the original state. I created another callback for the decline action. 

```Typescript
  /**
   * Called when users decline a palette change.
   */
  paletteChangeDecline() {
    if (this.paletteConfimationDialog) {
      this.paletteConfimationDialog.close();
    }
    if (this.paletteSelect) {
      this.paletteSelect.value = this.selectedPalette;
      this.requestedPalette = this.selectedPalette;
    }
  }
```

File: app-creator/client/src/widgets/dropzone-widget/dropzone-widget.ts
Comment (joannalcy) [reply app-creator/client/src/widgets/dropzone-widget/dropzone-widget.ts:145]: Yes I totally missed the collapsed part, my bad. :)

File: app-creator/client/src/widgets/story-board/story-board.ts
Comment (joannalcy) [inline app-creator/client/src/widgets/story-board/story-board.ts]: Nit: declinePaletteChange might be more natural?

**Review by joannalcy — APPROVED** [review]

File: app-creator/client/src/utils/helpers.ts
Comment (sufyanAbbasi) [reply app-creator/client/src/utils/helpers.ts:69]: This is awesome!

File: app-creator/client/src/redux/actions.ts
Comment (sufyanAbbasi) [inline app-creator/client/src/redux/actions.ts]: When I think of setting the palette, I think of setting the actual `PaletteProperties` and not just a palette name string. I would suggest refactoring this method (and the other methods on the callstack) to use the `PaletteProperties` directly. This will make it easier to support user provided palettes, since the user interface component can emit an entire palette instead of the simple string. 

I would also recommend renaming:
`Palette` --> `PaletteNames`
`PaletteProperties` --> `Palette`

File: app-creator/client/src/widgets/story-board/story-board.ts
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/story-board/story-board.ts]: nit: lowercase `Re-render`.

File: app-creator/client/src/widgets/story-board/story-board.ts
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/story-board/story-board.ts]: Is there a reason why we want to rerender the interface when this parameter changes? Since it's not being used in the rendering logic, it makes sense to me if this is moved up to the object level.

In fact, I believe we can completely remove this parameter altogether since the `select` should always point to the currently selected template anyway. What we could do instead is on `select` change, check that the new select mismatches the currently selected one. If so, trigger the dialog open, else do nothing. If confirmed, trigger `setPalette`, otherwise set the select back to `this.selectedPalette`. This will trigger the `onchange` again, but because there's no diff, it doesn't trigger the dialog.

File: app-creator/client/src/widgets/story-board/story-board.ts
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/story-board/story-board.ts]: It's better to have a proper label next to the select instead of using a disabled option. It also means you can set the default option to the one currently selected on the template!

File: app-creator/client/src/widgets/template-wizard/template-wizard.ts
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/template-wizard/template-wizard.ts]: nit: `Returns an array of template cards.`

File: app-creator/client/src/widgets/template-wizard/template-wizard.ts
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/template-wizard/template-wizard.ts]: nit: `getTextInput`, `getSelectInput`, and `getInputHeader` imply that they already exist. Perhaps these should be `make*`?

File: app-creator/client/src/widgets/template-wizard/template-wizard.ts
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/template-wizard/template-wizard.ts]: Buttons should have inverted color schemes, for better color contrast (`color` for background color and `backgroundColor` for `color`).

File: app-creator/client/src/widgets/template-wizard/template-wizard.ts
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/template-wizard/template-wizard.ts]: If this is true, we should explicitly set the `backgroundColor` to transparent in this method in an `else`. Also below for `panel` `color`.

File: app-creator/client/src/widgets/template-wizard/template-wizard.ts
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/template-wizard/template-wizard.ts]: It looks like we're applying `color` and `backgroundColor` to `map` elements, which is undefined. Should we move this to the top of the loop and put all the other logic in the `else`?

File: app-creator/client/src/widgets/template-wizard/template-wizard.ts
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/template-wizard/template-wizard.ts]: Same as above, have this be the palette properties rather than the color string.

File: app-creator/client/src/widgets/template-wizard/template-wizard.ts
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/template-wizard/template-wizard.ts:480]: Will changes from PR 19 be merged into this PR?

File: app-creator/client/src/widgets/template-wizard/test/template-wizard_test.ts
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/template-wizard/test/template-wizard_test.ts]: There are a few more tests that are useful here. For example, that the continue button is disabled while certain fields are blank, and whether the backgroundColor/color style logic is applied correctly to a template. I would recommend adding tests for any logic for which you need to guarantee something to be true for the rest of the app creator to work correctly, particularly around the template JSON.

**Review by sufyanAbbasi — CHANGES_REQUESTED** [review]

File: app-creator/client/src/widgets/palette-picker/palette-picker.ts
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/palette-picker/palette-picker.ts]: Since we're not planning to have this editable for now, let's mark it as `readonly`.

File: app-creator/client/src/redux/reducer.ts
Comment (joannalcy) [inline app-creator/client/src/redux/reducer.ts]: Are all events guaranteed to have a non empty `action.payload.value`? If not, eventType will be set to none, which might cause unwanted behaviors later on.

File: app-creator/client/src/widgets/palette-picker/test/palette-picker_test.ts
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/palette-picker/test/palette-picker_test.ts:2]: Let's add a test simulating a palette selection and that the event fires correctly and with the right data!

File: app-creator/client/src/redux/reducer.ts
Comment (joannalcy) [inline app-creator/client/src/redux/reducer.ts]: Use Object.assign to be consistent with the other cases below.

**Review by sufyanAbbasi — CHANGES_REQUESTED** [review]
The `PalettePicker` component looks really good! Using it as a store for palette information keeps things nice and modularized.

It looks like there are some opportunities to improve the tests, I will approve after you get those tests sorted out!

File: app-creator/client/src/redux/reducer.ts
Comment (joannalcy) [inline app-creator/client/src/redux/reducer.ts:242]: What's the difference between using either widgethIds?

File: app-creator/client/src/redux/reducer.ts
Comment (joannalcy) [inline app-creator/client/src/redux/reducer.ts]: Nit: widgetId

File: app-creator/client/src/widgets/palette-picker/palette-picker.ts
Comment (joannalcy) [inline app-creator/client/src/widgets/palette-picker/palette-picker.ts]: Color hexes should be lowercased: [go/htmlcssstyle#Capitalization](go/htmlcssstyle#Capitalization)

File: app-creator/client/src/widgets/template-wizard/test/template-wizard_test.ts
Comment (joannalcy) [reply app-creator/client/src/widgets/template-wizard/test/template-wizard_test.ts]: Agreed that we should add unit tests to make sure it works properly :)

File: app-creator/client/src/redux/reducer.ts
Comment (mibrah42) [reply app-creator/client/src/redux/reducer.ts]: The `value` argument is mandatory in the action and has a boolean type. We can also default it to true. 

```Typescript
export const setEventType = (
  eventType: EventType,
  value: boolean = true
): SetEventType => {
  return {
    type: SET_EVENT_TYPE,
    payload: {
      eventType,
      value,
    },
  };
};
```

File: app-creator/client/src/redux/reducer.ts
Comment (mibrah42) [reply app-creator/client/src/redux/reducer.ts:242]: `SET_SELECTED_TEMPLATE` is called to reset the template with all its state. So when we select new templates from the templates tab, or when we import a new template. In addition it is also called when changing palettes. However, in the case of palette change, we do not want to reset the widgetIDs.

File: app-creator/client/src/utils/helpers.ts
Comment (sufyanAbbasi) [inline app-creator/client/src/utils/helpers.ts]: I'll consolidate a few comments in one here:
- You can remove this condition if you move the `Array.isArray` condition up top.
- Do we not care if `key === WIDGET_REF` for the array case?
- Also, we don't apply `deepCloneTemplate` to the elements of the array (of widgets?), so they will not be cloned. Is that the expected behavior? If so, leave a comment about it and/or write a test for it.

File: app-creator/client/src/utils/test/helpers_test.ts
Comment (sufyanAbbasi) [inline app-creator/client/src/utils/test/helpers_test.ts:30]: Re:comment above, we'll want a test to check that the elements of the children aren't referring to the same objects.

**Review by sufyanAbbasi — APPROVED** [review]
I suggested only a few minor changes to make to the `deepClone` method. Once those are resolved, go ahead and merge!

File: app-creator/client/src/utils/helpers.ts
Comment (mibrah42) [reply app-creator/client/src/utils/helpers.ts]: - The key `WIDGET_REF` only stores a reference to an HTMLElement so there is no need to check it for the array case. 
- Our widgets are stored in an object so the only arrays we have contain primitive data types and thus cloning the array gives us new references. Nonetheless, we could call deepCloneTemplate again on each child.

File: app-creator/client/src/utils/test/helpers_test.ts
Comment (mibrah42) [reply app-creator/client/src/utils/test/helpers_test.ts:30]: In the Redux store, all array elements contain primitive data types, so checking if the array reference itself is different should be sufficient.

**Pull request merged**
